src/LeaseManager.h

```diff
@@ -0,0 +1,157 @@
+/* Copyright (c) 2014-2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef RAMCLOUD_LEASEMANAGER_H
+#define RAMCLOUD_LEASEMANAGER_H
+
+#include <set>
+#include <unordered_map>
+
+#include "Common.h"
+#include "CoordinatorClusterClock.h"
+#include "WireFormat.h"
+
+namespace RAMCloud {
+
+/**
+ * The LeaseManager, which resides on the coordinator, manages and acts as the
+ * central authority on the liveness of a client lease.  Client leases are used
+ * to track the liveness of any given client.  When a client holds a valid
+ * lease, it is considered active and various pieces of state (e.g.
+ * linearizability information) must be kept throughout the cluster.
+ *
+ * Clients must ensure they have an valid lease by periodically contacting this
+ * module.  Servers in their part must contact this module to check the validity
+ * of client leases.
+ */
+class LeaseManager {
+  PUBLIC:
+    explicit LeaseManager(Context *context);
+    WireFormat::ClientLease getLeaseInfo(uint64_t leaseId);
+    void recover();
+    WireFormat::ClientLease renewLease(uint64_t leaseId);
+    void startUpdaters();
+
+  PRIVATE:
+    /**
+     * The LeasePreallocator is periodically invoked to allocate a leaseId on
+     * external storage.  The goal is that this preallocator will work ahead
+     * of the issued leases so that a client does not have to wait for an
+     * external storage operation to complete a new lease request.  Every
+     * invocation of the preallocator should ensure that maxAllocatedLeaseId
+     * runs ahead of lastIssuedLeaseId by the PREALLOCATION_LIMIT.  This batch
+     * allocation process only blocks during each individual allocation; other
+     * operations like issuing leases can be safely interleaved.
+     */
+    class LeasePreallocator : public WorkerTimer {
+      public:
+        explicit LeasePreallocator(Context* context,
+                                   LeaseManager* leaseManager);
+        virtual void handleTimerEvent();
+
+        LeaseManager* leaseManager;
+      private:
+        DISALLOW_COPY_AND_ASSIGN(LeasePreallocator);
+    };
+
+    /**
+     * The LeaseCleaner periodically wakes up to expire leases.  Every time it
```

Comment:
A lot of this documentation feels like internal documentation; I'm wondering if you could get by here with just the first sentence.